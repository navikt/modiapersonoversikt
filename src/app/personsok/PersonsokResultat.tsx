import * as React from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../redux/reducers';
import { isFailedPosting, isFinishedPosting, isNotStartedPosting, isPosting } from '../../rest/utils/postResource';
import NavFrontendSpinner from 'nav-frontend-spinner';
import { AdresseCelle, BostedCelle, IdentCelle, NavnCelle } from './PersonsokResultatElementer';
import { ClickableTable } from '../../utils/table/ClickableTable';
import { AlertStripeAdvarsel, AlertStripeInfo } from 'nav-frontend-alertstriper';
import { setNyBrukerIPath } from '../routes/routing';
import { RouteComponentProps, withRouter } from 'react-router';

interface Props extends RouteComponentProps {
    onClose: () => void;
}

function PersonsokResultat(props: Props) {
    const personsokResource = useSelector((state: AppState) => state.restResources.personsok);

    if (isNotStartedPosting(personsokResource)) {
        return null;
    }

    if (isPosting(personsokResource)) {
        return <NavFrontendSpinner />;
    }

    if (isFailedPosting(personsokResource)) {
        return <AlertStripeAdvarsel>{personsokResource.error}</AlertStripeAdvarsel>;
    }

    if (!isFinishedPosting(personsokResource)) {
        return <AlertStripeAdvarsel>Noe gikk galt</AlertStripeAdvarsel>;
    }

    const tittelRekke = ['Fødselsnummer', 'Navn', 'Adresser', 'Bosted'];
    const response = personsokResource.response;

    if (response.length === 0) {
        return <AlertStripeInfo>Søket ga ingen treff</AlertStripeInfo>;
    }

    const tableEntries = response.map(linje => [
        <IdentCelle ident={linje.ident} />,
        <NavnCelle navn={linje.navn} status={linje.status} />,
        <AdresseCelle response={linje} />,
        <BostedCelle brukerinfo={linje.brukerinfo} />
    ]);
    const handlers = response.map(linje => () => {
        props.onClose();
        setNyBrukerIPath(props.history, linje.ident.ident);
    });

    return <ClickableTable tittelRekke={tittelRekke} rows={tableEntries} rowsOnClickHandlers={handlers} />;
}

export default withRouter(PersonsokResultat);
