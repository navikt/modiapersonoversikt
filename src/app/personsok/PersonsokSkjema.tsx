import * as React from 'react';
import { FormEvent, useState } from 'react';
import { PersonsokRequest } from '../../models/person/personsok';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../redux/reducers';
import PersonsokSkjemaElementer from './PersonsokSkjemaElementer';
import { ValideringsResultat } from '../../utils/forms/FormValidator';
import {
    getValidPersonSokState,
    personsokSkjemaHarNokInformasjonTilÅGjøreSøk,
    validerPersonsokSkjema
} from './personsokValidator';
import { AlertStripeInfo } from 'nav-frontend-alertstriper';
import { loggPersonsok } from './loggPersonsok';

export type PersonSokFormState = {
    fornavn: string;
    etternavn: string;
    gatenavn: string;
    husnummer: string;
    husbokstav: string;
    postnummer: string;
    kontonummer: string;
    kommunenummer: string;
    fodselsdatoFra?: string;
    fodselsdatoTil?: string;
    alderFra: string;
    alderTil: string;
    kjonn: string;
};

export interface PersonsokSkjemaProps {
    state: PersonSokFormState;
    actions: {
        settFornavn(fornavn: string): void;
        settEtternavn(etternavn: string): void;
        settGatenavn(gatenavn: string): void;
        settHusnummer(husnummer: string): void;
        settHusbokstav(husbokstav: string): void;
        settPostnummer(postnummer: string): void;
        settKontonummer(kontonummer: string): void;
        settKommunenummer(kommunenummer: string): void;
        settFodselsdatoFra(fodselsdatoFra: string | undefined): void;
        settFodselsdatoTil(fodselsdatoTil: string | undefined): void;
        settAlderFra(alderFra: string): void;
        settAlderTil(alderTil: string): void;
        settKjonn(kjonn: string): void;
    };
    valideringsResultat: ValideringsResultat<PersonSokFormState>;
}

function stringToNumber(input: string): number | undefined {
    if (input.length === 0) {
        return undefined;
    }
    return parseInt(input);
}

function emptyString(input: string): string | undefined {
    if (input.length === 0) {
        return undefined;
    }
    return input;
}

function lagRequest(form: PersonsokSkjemaProps): PersonsokRequest {
    return {
        fornavn: emptyString(form.state.fornavn),
        etternavn: emptyString(form.state.etternavn),
        gatenavn: emptyString(form.state.gatenavn),
        husnummer: stringToNumber(form.state.husnummer),
        husbokstav: emptyString(form.state.husbokstav),
        postnummer: emptyString(form.state.postnummer),
        kontonummer: emptyString(form.state.kontonummer),
        kommunenummer: emptyString(form.state.kommunenummer),
        fodselsdatoFra: form.state.fodselsdatoFra,
        fodselsdatoTil: form.state.fodselsdatoTil,
        alderFra: stringToNumber(form.state.alderFra),
        alderTil: stringToNumber(form.state.alderTil),
        kjonn: emptyString(form.state.kjonn)
    };
}

function PersonsokSkjema() {
    const dispatch = useDispatch();
    const personsokResource = useSelector((state: AppState) => state.restResources.personsok);
    const [fornavn, settFornavn] = useState<string>('');
    const [etternavn, settEtternavn] = useState<string>('');
    const [gatenavn, settGatenavn] = useState<string>('');
    const [husnummer, settHusnummer] = useState<string>('');
    const [husbokstav, settHusbokstav] = useState<string>('');
    const [postnummer, settPostnummer] = useState<string>('');
    const [kontonummer, settKontonummer] = useState<string>('');
    const [kommunenummer, settKommunenummer] = useState<string>('');
    const [fodselsdatoFra, settFodselsdatoFra] = useState<string | undefined>(undefined);
    const [fodselsdatoTil, settFodselsdatoTil] = useState<string | undefined>(undefined);
    const [alderFra, settAlderFra] = useState<string>('');
    const [alderTil, settAlderTil] = useState<string>('');
    const [kjonn, settKjonn] = useState<string>('');
    const [valideringsResultat, settValideringsresultat] = useState<ValideringsResultat<PersonSokFormState>>(
        getValidPersonSokState()
    );
    const [minimumsKriterierOppfylt, setMinimumsKriterierOppfylt] = useState<boolean>(true);

    const formState: PersonsokSkjemaProps = {
        state: {
            fornavn,
            etternavn,
            gatenavn,
            husnummer,
            husbokstav,
            postnummer,
            kontonummer,
            kommunenummer,
            fodselsdatoFra,
            fodselsdatoTil,
            alderFra,
            alderTil,
            kjonn
        },
        actions: {
            settFornavn,
            settEtternavn,
            settGatenavn,
            settHusnummer,
            settHusbokstav,
            settPostnummer,
            settKontonummer,
            settKommunenummer,
            settFodselsdatoFra,
            settFodselsdatoTil,
            settAlderFra,
            settAlderTil,
            settKjonn
        },
        valideringsResultat: valideringsResultat
    };

    const submitHandler = (event: FormEvent) => {
        event.preventDefault();
        if (!personsokSkjemaHarNokInformasjonTilÅGjøreSøk(formState.state)) {
            setMinimumsKriterierOppfylt(false);
            return;
        }
        setMinimumsKriterierOppfylt(true);
        const valideringsResultat = validerPersonsokSkjema(formState.state);
        if (valideringsResultat.formErGyldig) {
            settValideringsresultat(getValidPersonSokState());
            const request: PersonsokRequest = lagRequest(formState);
            loggPersonsok(request);
            dispatch(personsokResource.actions.post(request));
        } else {
            settValideringsresultat(valideringsResultat);
        }
    };

    const resetHandler = (event: FormEvent) => {
        settFornavn('');
        settEtternavn('');
        settGatenavn('');
        settHusnummer('');
        settHusbokstav('');
        settPostnummer('');
        settKontonummer('');
        settKommunenummer('');
        settFodselsdatoFra(undefined);
        settFodselsdatoTil(undefined);
        settAlderFra('');
        settAlderTil('');
        settKjonn('');
    };

    return (
        <form onSubmit={submitHandler} onReset={resetHandler}>
            <PersonsokSkjemaElementer form={formState} />
            {!minimumsKriterierOppfylt && (
                <AlertStripeInfo>
                    <span role="alert">Du må minimum fylle inn navn, adresse eller kontonummer for å gjøre søk</span>
                </AlertStripeInfo>
            )}
        </form>
    );
}

export default PersonsokSkjema;
