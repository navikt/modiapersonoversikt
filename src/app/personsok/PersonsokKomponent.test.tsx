import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TestProvider from '../../test/Testprovider';
import PersonsokResultat from './PersonsokResultat';
import PersonsokSkjema from './PersonsokSkjema';
import { setupReactQueryMocks } from '../../test/testStore';

test('viser personsÃ¸k-komponent', () => {
    setupReactQueryMocks();
    const personsokKomponentV3 = renderer.create(
        <TestProvider>
            <>
                <PersonsokSkjema setPosting={() => {}} setResponse={() => {}} />
                <PersonsokResultat posting={false} response={undefined} onClose={() => {}} />
            </>
        </TestProvider>
    );

    expect(personsokKomponentV3.toJSON()).toMatchSnapshot();
});
