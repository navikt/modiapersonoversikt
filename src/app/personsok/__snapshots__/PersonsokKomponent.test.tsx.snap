// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viser personsøk-komponent 1`] = `
.c5 {
  margin-right: 0.5em;
  margin-bottom: 1em;
  line-height: 1.375rem;
}

.c5 .nav-datovelger__kalenderPortal {
  position: fixed;
  top: 0px;
  left: 0px;
  -webkit-transform: translateX(-85%);
  -ms-transform: translateX(-85%);
  transform: translateX(-85%);
}

.c6 {
  margin-bottom: 0.5em;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  position: relative;
  border: none;
  padding: 0;
  border-radius: .35rem;
  cursor: pointer;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #0067c5;
}

.c8:after {
  border-bottom: 1px #b7b1a9 solid;
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
}

.c8:focus {
  outline: none;
  box-shadow: 0 0 0 0.1875rem #254b6d;
}

.c8:hover {
  opacity: 0.8;
}

.c8:hover:after {
  content: '';
  border-color: #0067c5;
}

.c0 {
  padding: .5rem;
}

.c0 .skjemaelement {
  margin-bottom: 0.5rem;
}

.c0 .skjemaelement__label {
  margin-bottom: 0rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.5rem;
}

.c1 > *:first-child {
  margin-right: 10rem;
}

.c7 {
  padding-top: .5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 .skjemaelement {
  padding-right: 0.5em;
}

<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <article
    className="c0"
  >
    <section
      className="c1"
    >
      <section
        aria-label="Søkekriterier"
      >
        <h2
          className="typo-systemtittel"
        >
          Søkekriterier
        </h2>
        <div
          className="c2"
        >
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="fornavn"
            >
              Fornavn (Fonetisk søk)
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--xl"
              id="fornavn"
              name="fornavn"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="etternavn"
            >
              Etternavn (Fonetisk søk)
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--xl"
              id="etternavn"
              name="etternavn"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="c2"
        >
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="gatenavn"
            >
              Gatenavn
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--l"
              id="gatenavn"
              name="gatenavn"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="husnummer"
            >
              Husnummer
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--m"
              id="husnummer"
              name="husnummer"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="husbokstav"
            >
              Husbokstav
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--m"
              id="husbokstav"
              name="husbokstav"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="skjemaelement"
        >
          <label
            className="skjemaelement__label"
            htmlFor="postnummer"
          >
            Postnummer
          </label>
          <input
            aria-invalid={false}
            className="skjemaelement__input input--m"
            id="postnummer"
            name="postnummer"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="skjemaelement"
        >
          <label
            className="skjemaelement__label"
            htmlFor="kontonummer"
          >
            Kontonummer (Norske nummer)
          </label>
          <input
            aria-invalid={false}
            className="skjemaelement__input input--l"
            id="kontonummer"
            name="kontonummer"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="skjemaelement"
        >
          <label
            className="skjemaelement__label"
            htmlFor="utenlandskID"
          >
            Utenlandsk ID 
            <div
              className="hjelpetekst"
            >
              <button
                aria-controls="Helt tilfeldig ID"
                aria-expanded={false}
                aria-haspopup="dialog"
                className="hjelpetekst__apneknapp"
                id="Helt tilfeldig ID"
                onClick={[Function]}
                title="Hjelp"
                type="button"
              >
                <svg
                  className="hjelpetekst__ikon"
                  focusable="false"
                  height={32}
                  kind="help-circle"
                  viewBox="0 0 18.2 18.2"
                  width={32}
                >
                  <path
                    d="M9.1 0C4.1 0 0 4.1 0 9.1s4.1 9.1 9.1 9.1 9.1-4.1 9.1-9.1S14.1 0 9.1 0zm0 17.2C4.6 17.2 1 13.6 1 9.1S4.6 1 9.1 1s8.1 3.6 8.1 8.1-3.6 8.1-8.1 8.1z"
                  />
                  <circle
                    cx="9.1"
                    cy="13.8"
                    r=".9"
                  />
                  <path
                    d="M9.1 11.5c-.3 0-.5-.2-.5-.5V8.6c0-.3.2-.5.5-.5 1 0 1.9-.8 1.9-1.9s-.8-1.9-1.9-1.9c-1 0-1.9.8-1.9 1.9 0 .3-.2.5-.5.5s-.5-.2-.5-.4c0-1.6 1.3-2.9 2.9-2.9S12 4.7 12 6.3c0 1.4-1 2.6-2.4 2.8V11c0 .3-.2.5-.5.5z"
                  />
                </svg>
                <span
                  className="sr-only"
                >
                  Hjelp
                </span>
              </button>
              <div
                className="popover popover--controlled popover--hidden popover--over"
                id="Helt tilfeldig ID"
                onClick={[Function]}
                role="tooltip"
                style={
                  Object {
                    "transform": "translate3d(0px, 0px, 0)",
                  }
                }
                tabIndex={0}
              >
                <div
                  className="popover__content-inner"
                >
                  <div
                    className="hjelpetekst__innhold"
                  >
                    Husk å inkludere alle tegn. Eksempel: 010101-12345
                  </div>
                </div>
                <span
                  className="popover__pil"
                  style={
                    Object {
                      "left": undefined,
                    }
                  }
                />
              </div>
            </div>
          </label>
          <input
            aria-invalid={false}
            className="skjemaelement__input input--l"
            id="utenlandskID"
            name="utenlandskID"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </section>
      <section
        aria-label="Begrens søket"
      >
        <h2
          className="typo-systemtittel"
        >
          Begrens søket
        </h2>
        <div
          className="skjemaelement"
        >
          <label
            className="skjemaelement__label"
            htmlFor="kommunenummer"
          >
            Bosted
          </label>
          <input
            aria-invalid={false}
            className="skjemaelement__input input--m"
            id="kommunenummer"
            name="kommunenummer"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="c2"
        >
          <div
            className="c3"
          >
            <div
              className="c4"
            >
              <div
                className="c5"
              >
                <label
                  className="c6 skjemaelement__label"
                  htmlFor="fodselsdatoFra"
                >
                  <p
                    className="typo-normal"
                  >
                    Fødselsdato fra
                  </p>
                </label>
                <div
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                >
                  <div
                    className="nav-datovelger"
                  >
                    <div
                      className="nav-datovelger__inputContainer"
                    >
                      <input
                        aria-invalid={false}
                        autoComplete="off"
                        autoCorrect="off"
                        className="nav-datovelger__input"
                        id="fodselsdatoFra"
                        maxLength={10}
                        name="fodselsdatoFra"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        pattern="\\\\d{2}.\\\\d{2}.\\\\d{4}"
                        type="text"
                        value=""
                      />
                      <button
                        aria-expanded={false}
                        aria-label="Kalender"
                        className="nav-datovelger__kalenderknapp"
                        onClick={[Function]}
                        role="button"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          height={16}
                          role="presentation"
                          viewBox="0 0 18 18"
                          width={16}
                        >
                          <title>
                            Kalenderikon
                          </title>
                          <g
                            fill="none"
                            fillRule="evenodd"
                            stroke="#0067C5"
                          >
                            <path
                              d="M4 2.667H1.333v14h15.334v-14H14"
                            />
                            <path
                              d="M4 1.333h2V4H4zm8 0h2V4h-2zM6 2h6M1.333 6h15.334"
                            />
                          </g>
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="c5"
              >
                <label
                  className="c6 skjemaelement__label"
                  htmlFor="fodselsdatoTil"
                >
                  <p
                    className="typo-normal"
                  >
                    Fødselsdato til
                  </p>
                </label>
                <div
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                >
                  <div
                    className="nav-datovelger"
                  >
                    <div
                      className="nav-datovelger__inputContainer"
                    >
                      <input
                        aria-invalid={false}
                        autoComplete="off"
                        autoCorrect="off"
                        className="nav-datovelger__input"
                        id="fodselsdatoTil"
                        maxLength={10}
                        name="fodselsdatoTil"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        pattern="\\\\d{2}.\\\\d{2}.\\\\d{4}"
                        type="text"
                        value=""
                      />
                      <button
                        aria-expanded={false}
                        aria-label="Kalender"
                        className="nav-datovelger__kalenderknapp"
                        onClick={[Function]}
                        role="button"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          height={16}
                          role="presentation"
                          viewBox="0 0 18 18"
                          width={16}
                        >
                          <title>
                            Kalenderikon
                          </title>
                          <g
                            fill="none"
                            fillRule="evenodd"
                            stroke="#0067C5"
                          >
                            <path
                              d="M4 2.667H1.333v14h15.334v-14H14"
                            />
                            <path
                              d="M4 1.333h2V4H4zm8 0h2V4h-2zM6 2h6M1.333 6h15.334"
                            />
                          </g>
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div />
          </div>
        </div>
        <div
          className="c2"
        >
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="alderFra"
            >
              Alder fra
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--m"
              id="alderFra"
              name="alderFra"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="skjemaelement"
          >
            <label
              className="skjemaelement__label"
              htmlFor="alderTil"
            >
              Alder til
            </label>
            <input
              aria-invalid={false}
              className="skjemaelement__input input--m"
              id="alderTil"
              name="alderTil"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="skjemaelement"
        >
          <label
            className="skjemaelement__label"
            htmlFor="kjonn"
          >
            Kjønn
          </label>
          <div
            className="selectContainer input--fullbredde"
          >
            <select
              aria-invalid={false}
              className="skjemaelement__input"
              disabled={false}
              id="kjonn"
              name="kjonn"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            >
              <option
                value=""
              >
                Velg Kjønn
              </option>
              <option
                value="K"
              >
                K - Kvinne
              </option>
              <option
                value="M"
              >
                M - Mann
              </option>
            </select>
          </div>
        </div>
      </section>
    </section>
    <a
      className="lenke"
      href="https://drek.nais.preprod.local/?fornavn=&etternavn=&foedselsdato=&kjoenn="
      rel="noopener noreferrer"
      target="_blank"
    >
      <p
        className="typo-normal"
      >
        Rekvirer D-nummer fra Folkeregisteret
      </p>
    </a>
    <div
      className="c7"
    >
      <button
        className="knapp knapp--hoved"
        disabled={false}
        type="submit"
      >
        Søk
      </button>
      <button
        className="c8"
        type="reset"
      >
        Nullstill
      </button>
    </div>
  </article>
</form>
`;
